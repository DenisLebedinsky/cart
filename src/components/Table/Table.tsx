import * as React from "react";

import TableRow from "../../containers/TableRowContainer";
import TableHeaderCell from "../TableHeaderCell";
import { stableSort, getSorting } from "../../utils/sort";

import css from "./_Table.module.scss";

import { Product } from "../../redux/reducers/cart/types";
import { orderBy, TableProps, TableState } from "./Table.type";

class Table extends React.Component<TableProps, TableState> {
  public state: Readonly<TableState> = {
    order: "DESC",
    orderBy: null
  };

  public handleClick = (column: orderBy) => () => {
    let { order } = this.state;

    if (column === this.state.orderBy && order === "DESC") {
      order = "ASC";
    } else {
      order = "DESC";
    }

    this.setState({ order, orderBy: column });
  };

  public getClassName = (column: orderBy) => {
    const { orderBy, order } = this.state;

    if (orderBy === column) {
      return order === "DESC" ? css.desc : css.asc;
    }

    return "";
  };

  public renderItems = () => {
    const { orderBy, order } = this.state;
    let { items } = this.props;

    if (orderBy) {
      items = stableSort(items, getSorting(order, orderBy));
    }

    return items.map((item: Product) => (
      <TableRow
        key={item.ProductId}
        id={item.ProductId}
        name={item.Name}
        quantity={item.Quantity}
        price={item.Price}
      />
    ));
  };

  public render() {
    return (
      <div className={css.table}>
        <div className={css.header}>
          <TableHeaderCell
            columnName="Product"
            className={this.getClassName("Name")}
            onClick={this.handleClick("Name")}
          />
          <TableHeaderCell
            columnName="Quantity"
            className={this.getClassName("Quantity")}
            onClick={this.handleClick("Quantity")}
          />
          <TableHeaderCell
            columnName="Price"
            className={this.getClassName("Price")}
            onClick={this.handleClick("Price")}
          />
          <TableHeaderCell />
        </div>
        {this.renderItems()}
      </div>
    );
  }
}

export default Table;
