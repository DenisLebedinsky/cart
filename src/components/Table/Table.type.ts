import { Product } from "../../redux/reducers/cart/types";

export interface TableState {
  orderBy: orderBy;
  order: order;
}

export interface TableProps {
  items: Product[];
}

export type order = "ASC" | "DESC";

export type orderBy = keyof PartialProduct | null;

type PartialProduct = Pick<Product, "Name" | "Quantity" | "Price">;
