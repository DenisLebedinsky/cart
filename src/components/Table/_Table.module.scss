.table {
  margin: 20px auto;
  background-color: #F5F5F5;
  min-width: 510px;
  max-width: 1200px;
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.4);

  .header {
    display: flex;

    > div {
      padding: 10px;

      > p {
        text-align: center;
        font-size: 18px;
        font-weight: bold;
        cursor: pointer;
      }

      &:first-child {
        flex: 4;
      }

      &:nth-child(2) {
        flex: 3;
      }

      &:nth-child(3) {
        flex: 3;
      }

      &:last-child {
        flex: 0 0 auto;
        width: 120px;
      }

      &.asc {
        display: flex;
        justify-content: center;
        align-items: center;

        &:after {
          content: "▲";
          margin-left: 10px;
        }
      }

      &.desc {
        display: flex;
        justify-content: center;
        align-items: center;

        &:after {
          content: "▼";
          margin-left: 10px;
        }
      }
    }
  }

  > div {
    border-bottom: 1px solid #000000;
    border-left: 1px solid #000000;
    border-right: 1px solid #000000;

    &:first-child {
      border-top: 1px solid #000000;
    }
  }
}
