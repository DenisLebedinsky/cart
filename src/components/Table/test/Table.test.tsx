import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import Table from "../Table";

describe("Table", () => {
  let wrapper: ShallowWrapper<Table>;

  const items: any[] = [];

  beforeEach(() => {
    wrapper = shallow(<Table items={items} />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should return desc when orderBy equals to column and order is 'DESC'", () => {
    const column = "Name";
    const order = "DESC";

    const instance = wrapper.instance() as Table;
    instance.setState({ orderBy: column, order });

    const result = instance.getClassName(column);

    expect(result).toBe("desc");
  });

  it("should return asc when orderBy equals to column and order is 'ASC'", () => {
    const column = "Name";
    const order = "ASC";

    const instance = wrapper.instance() as Table;
    instance.setState({ orderBy: column, order });

    const result = instance.getClassName(column);

    expect(result).toBe("asc");
  });

  it("should set state order to 'DESC' and orderBy to 'Name'", () => {
    const column = "Name";
    const instance = wrapper.instance() as Table;

    instance.handleClick(column)();

    expect(instance.state.order).toBe("DESC");
  });

  it("should set state order to 'ASC' and orderBy to 'Name'", () => {
    const column = "Name";
    const instance = wrapper.instance() as Table;

    instance.handleClick(column)();
    instance.handleClick(column)();

    expect(instance.state.order).toBe("ASC");
  });
});
