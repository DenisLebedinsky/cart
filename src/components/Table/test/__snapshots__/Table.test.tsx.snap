// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match the snapshot 1`] = `
<div
  className="table"
>
  <div
    className="header"
  >
    <TableHeaderCell
      className=""
      columnName="Product"
      onClick={[Function]}
    />
    <TableHeaderCell
      className=""
      columnName="Quantity"
      onClick={[Function]}
    />
    <TableHeaderCell
      className=""
      columnName="Price"
      onClick={[Function]}
    />
    <TableHeaderCell />
  </div>
</div>
`;
